In [ ]:
from urllib.parse import urlparse
from urllib.parse import urldefrag
from urllib.request import urlopen
from urllib.parse import urljoin
from bs4 import BeautifulSoup
from file_storage import FileStorage
import time
import datetime

In [7]:
visited_urls = set()
parse_queue = []
bad_urls = []
url = 'https://simple.wikipedia.org/wiki/Main_Page'
parse_queue.append(url)

In [8]:
fs = FileStorage('storage')
cnt = 0

In [ ]:
%%time
for url in parse_queue:
    url = urldefrag(url).url
    parsed_url = urlparse(url)
    base_url = parsed_url.netloc + parsed_url.path
    if base_url in visited_urls:
        continue
    visited_urls.add(base_url)
    if cnt % 100 == 0:
        print('number of parsed urls = {}, time = {}'.format(cnt, datetime.datetime.now().time()))
    cnt += 1
    try:
        html = urlopen(url).read().decode('utf-8')
        fs.write(url, html)
        parser = BeautifulSoup(html)
        for link in parser.findAll('a'):
            next_url = urljoin(url, link.get('href'))
            next_url = urldefrag(next_url).url
            parsed_next = urlparse(next_url)
            netloc = parsed_next.netloc
            path = parsed_next.path
            key = netloc + path
            if key in visited_urls or (netloc != "simple.wikipedia.org" and netloc != ""):
                continue
            if (':' in path and ('Category:' not in path)) or (path.count(':') > 1) or (path[:5] != '/wiki') \
                or ('#' in next_url):
                bad_urls.append(next_url)
                continue
            parse_queue.append(next_url)
    except:
        print('parsing failed {}'.format(url))

number of parsed urls = 0, time = 01:29:39.011434
number of parsed urls = 100, time = 01:30:13.025576
number of parsed urls = 200, time = 01:30:45.673245
number of parsed urls = 300, time = 01:31:17.436115
number of parsed urls = 400, time = 01:31:45.995425
number of parsed urls = 500, time = 01:32:19.114870
number of parsed urls = 600, time = 01:32:52.470543
number of parsed urls = 700, time = 01:33:24.489581
number of parsed urls = 800, time = 01:33:56.611999
number of parsed urls = 900, time = 01:34:28.328873
number of parsed urls = 1000, time = 01:35:01.753495
number of parsed urls = 1100, time = 01:35:37.318735
number of parsed urls = 1200, time = 01:36:07.832191
number of parsed urls = 1300, time = 01:36:39.477499
number of parsed urls = 1400, time = 01:37:10.713117
number of parsed urls = 1500, time = 01:37:39.244328
number of parsed urls = 1600, time = 01:38:09.296704
number of parsed urls = 1700, time = 01:38:43.309615
number of parsed urls = 1800, time = 01:39:13.638089
numbe

number of parsed urls = 15400, time = 02:51:02.640694
number of parsed urls = 15500, time = 02:51:33.752202
number of parsed urls = 15600, time = 02:52:03.023468
number of parsed urls = 15700, time = 02:52:32.762178
number of parsed urls = 15800, time = 02:53:11.452406
number of parsed urls = 15900, time = 02:53:43.024621
number of parsed urls = 16000, time = 02:54:12.640508
number of parsed urls = 16100, time = 02:54:55.045831
number of parsed urls = 16200, time = 02:55:44.754705
number of parsed urls = 16300, time = 02:56:33.776996
number of parsed urls = 16400, time = 02:57:05.652735
number of parsed urls = 16500, time = 02:57:38.287066
number of parsed urls = 16600, time = 02:58:10.007343
number of parsed urls = 16700, time = 02:58:41.690315
number of parsed urls = 16800, time = 02:59:14.603559
number of parsed urls = 16900, time = 02:59:44.646877
number of parsed urls = 17000, time = 03:00:16.823014
number of parsed urls = 17100, time = 03:00:47.440608
number of parsed urls = 1720

number of parsed urls = 29200, time = 04:07:32.816602
number of parsed urls = 29300, time = 04:08:06.168202
number of parsed urls = 29400, time = 04:08:40.727983
number of parsed urls = 29500, time = 04:09:11.811058
number of parsed urls = 29600, time = 04:09:47.554972
number of parsed urls = 29700, time = 04:10:21.214468
number of parsed urls = 29800, time = 04:10:52.554452
number of parsed urls = 29900, time = 04:11:29.358109
number of parsed urls = 30000, time = 04:12:03.930325
number of parsed urls = 30100, time = 04:12:34.892207
number of parsed urls = 30200, time = 04:13:09.605103
number of parsed urls = 30300, time = 04:13:42.645467
number of parsed urls = 30400, time = 04:14:19.743775
number of parsed urls = 30500, time = 04:14:55.591102
number of parsed urls = 30600, time = 04:15:28.165545
number of parsed urls = 30700, time = 04:16:02.039956
number of parsed urls = 30800, time = 04:16:37.757261
number of parsed urls = 30900, time = 04:17:12.683962
number of parsed urls = 3100

number of parsed urls = 43400, time = 05:27:44.223983
number of parsed urls = 43500, time = 05:28:16.961813
number of parsed urls = 43600, time = 05:28:51.424731
number of parsed urls = 43700, time = 05:29:27.235406
number of parsed urls = 43800, time = 05:29:59.203444
number of parsed urls = 43900, time = 05:30:35.167903
number of parsed urls = 44000, time = 05:31:07.255488
number of parsed urls = 44100, time = 05:31:39.299127
number of parsed urls = 44200, time = 05:32:14.021497
number of parsed urls = 44300, time = 05:32:45.070157
number of parsed urls = 44400, time = 05:33:16.292319
number of parsed urls = 44500, time = 05:33:48.176899
number of parsed urls = 44600, time = 05:34:20.375558
number of parsed urls = 44700, time = 05:34:56.454244
number of parsed urls = 44800, time = 05:35:31.377538
number of parsed urls = 44900, time = 05:36:14.255741
number of parsed urls = 45000, time = 05:36:56.325225
number of parsed urls = 45100, time = 05:37:30.443363
number of parsed urls = 4520

number of parsed urls = 58600, time = 06:50:38.972446
number of parsed urls = 58700, time = 06:51:11.519933
number of parsed urls = 58800, time = 06:51:44.001759
number of parsed urls = 58900, time = 06:52:16.245943
number of parsed urls = 59000, time = 06:52:48.280855
number of parsed urls = 59100, time = 06:53:20.538887
number of parsed urls = 59200, time = 06:53:51.086092
number of parsed urls = 59300, time = 06:54:22.279546
number of parsed urls = 59400, time = 06:54:53.546013
number of parsed urls = 59500, time = 06:55:26.040080
number of parsed urls = 59600, time = 06:55:56.592095
number of parsed urls = 59700, time = 06:56:30.488848
number of parsed urls = 59800, time = 06:57:01.282061
number of parsed urls = 59900, time = 06:57:30.879414
number of parsed urls = 60000, time = 06:58:00.811467
number of parsed urls = 60100, time = 06:58:30.551491
number of parsed urls = 60200, time = 06:59:02.730017
number of parsed urls = 60300, time = 06:59:33.726158
number of parsed urls = 6040

number of parsed urls = 73400, time = 08:12:07.008699
number of parsed urls = 73500, time = 08:12:44.784487
number of parsed urls = 73600, time = 08:13:20.517822
parsing failed https://simple.wikipedia.org/wiki/Duchy_of_Swabia
number of parsed urls = 73700, time = 08:13:54.743714
number of parsed urls = 73800, time = 08:14:25.328019
number of parsed urls = 73900, time = 08:15:01.291288
number of parsed urls = 74000, time = 08:15:35.356379
number of parsed urls = 74100, time = 08:16:07.828344
number of parsed urls = 74200, time = 08:16:39.020252
number of parsed urls = 74300, time = 08:17:13.051955
number of parsed urls = 74400, time = 08:17:45.568427
number of parsed urls = 74500, time = 08:18:18.864800
number of parsed urls = 74600, time = 08:18:50.010403
number of parsed urls = 74700, time = 08:19:21.267849
number of parsed urls = 74800, time = 08:19:54.920508
number of parsed urls = 74900, time = 08:20:27.696636
number of parsed urls = 75000, time = 08:21:02.964956
number of parsed 

number of parsed urls = 87900, time = 09:33:13.622172
number of parsed urls = 88000, time = 09:33:45.088941
number of parsed urls = 88100, time = 09:34:18.558220
number of parsed urls = 88200, time = 09:34:58.622786
number of parsed urls = 88300, time = 09:35:37.729324
number of parsed urls = 88400, time = 09:36:18.343535
number of parsed urls = 88500, time = 09:36:52.447284
number of parsed urls = 88600, time = 09:37:27.015692
number of parsed urls = 88700, time = 09:38:00.574885
number of parsed urls = 88800, time = 09:38:44.379007
number of parsed urls = 88900, time = 09:39:31.976619
number of parsed urls = 89000, time = 09:40:05.361716
number of parsed urls = 89100, time = 09:40:37.164341
number of parsed urls = 89200, time = 09:41:09.417994
number of parsed urls = 89300, time = 09:41:42.875610
number of parsed urls = 89400, time = 09:42:17.660224
number of parsed urls = 89500, time = 09:42:51.748901
number of parsed urls = 89600, time = 09:43:23.816493
number of parsed urls = 8970

number of parsed urls = 97100, time = 10:26:38.060661
number of parsed urls = 97200, time = 10:27:13.194323
number of parsed urls = 97300, time = 10:27:47.059789
number of parsed urls = 97400, time = 10:28:17.553206
number of parsed urls = 97500, time = 10:28:50.856090
number of parsed urls = 97600, time = 10:29:24.297468
number of parsed urls = 97700, time = 10:29:57.845902
number of parsed urls = 97800, time = 10:30:31.612882
number of parsed urls = 97900, time = 10:31:05.824706
number of parsed urls = 98000, time = 10:31:40.140261
number of parsed urls = 98100, time = 10:32:13.201749
number of parsed urls = 98200, time = 10:32:46.181591
number of parsed urls = 98300, time = 10:33:18.604512
number of parsed urls = 98400, time = 10:33:50.533072
number of parsed urls = 98500, time = 10:34:23.563100
number of parsed urls = 98600, time = 10:34:57.208064
number of parsed urls = 98700, time = 10:35:29.753594
number of parsed urls = 98800, time = 10:36:05.137333
number of parsed urls = 9890

number of parsed urls = 109600, time = 11:37:12.239631
number of parsed urls = 109700, time = 11:37:44.169838
parsing failed https://simple.wikipedia.org/wiki/IPA_vowels_chart_with_audio
number of parsed urls = 109800, time = 11:38:17.933377
number of parsed urls = 109900, time = 11:38:52.539685
number of parsed urls = 110000, time = 11:39:29.641193
number of parsed urls = 110100, time = 11:40:03.014989
number of parsed urls = 110200, time = 11:40:38.092187
number of parsed urls = 110300, time = 11:41:14.143674
number of parsed urls = 110400, time = 11:41:47.931204
number of parsed urls = 110500, time = 11:42:22.528110
number of parsed urls = 110600, time = 11:42:54.340594
number of parsed urls = 110700, time = 11:43:29.728130
number of parsed urls = 110800, time = 11:44:03.667920
number of parsed urls = 110900, time = 11:44:36.872753
number of parsed urls = 111000, time = 11:45:08.160965
number of parsed urls = 111100, time = 11:45:39.285934
number of parsed urls = 111200, time = 11:4

number of parsed urls = 124200, time = 12:59:58.702146
number of parsed urls = 124300, time = 13:00:32.742933
number of parsed urls = 124400, time = 13:01:04.704370
number of parsed urls = 124500, time = 13:01:38.598834
number of parsed urls = 124600, time = 13:02:11.340303
number of parsed urls = 124700, time = 13:02:50.225264
number of parsed urls = 124800, time = 13:03:26.110004
number of parsed urls = 124900, time = 13:03:59.620952
number of parsed urls = 125000, time = 13:04:34.112284
parsing failed https://simple.wikipedia.org/wiki/Kingdom_of_Montenegro
number of parsed urls = 125100, time = 13:05:10.991317
number of parsed urls = 125200, time = 13:05:42.906271
parsing failed https://simple.wikipedia.org/wiki/Hero_of_Socialist_Labor
number of parsed urls = 125300, time = 13:06:19.048649
number of parsed urls = 125400, time = 13:06:51.201128
number of parsed urls = 125500, time = 13:07:27.181760
number of parsed urls = 125600, time = 13:08:01.619364
number of parsed urls = 125700,

number of parsed urls = 131200, time = 13:41:48.094160
number of parsed urls = 131300, time = 13:42:23.159332
number of parsed urls = 131400, time = 13:43:02.834902
number of parsed urls = 131500, time = 13:43:42.673422
number of parsed urls = 131600, time = 13:44:18.581355
number of parsed urls = 131700, time = 13:44:54.298966
number of parsed urls = 131800, time = 13:45:30.154885
number of parsed urls = 131900, time = 13:46:05.785325
number of parsed urls = 132000, time = 13:46:39.305043
number of parsed urls = 132100, time = 13:47:14.072559
number of parsed urls = 132200, time = 13:47:47.408842
number of parsed urls = 132300, time = 13:48:24.143909
number of parsed urls = 132400, time = 13:49:00.963122
number of parsed urls = 132500, time = 13:49:37.680862
parsing failed https://simple.wikipedia.org/wiki/Flag_of_Johor
number of parsed urls = 132600, time = 13:50:13.221268
number of parsed urls = 132700, time = 13:50:51.857383
number of parsed urls = 132800, time = 13:51:32.008717
nu

number of parsed urls = 142500, time = 14:49:18.500091
number of parsed urls = 142600, time = 14:49:54.775611
number of parsed urls = 142700, time = 14:50:29.204524
number of parsed urls = 142800, time = 14:51:02.658519
number of parsed urls = 142900, time = 14:51:35.862710
number of parsed urls = 143000, time = 14:52:12.233248
number of parsed urls = 143100, time = 14:52:46.379374
number of parsed urls = 143200, time = 14:53:21.421562
number of parsed urls = 143300, time = 14:53:54.753839
number of parsed urls = 143400, time = 14:54:29.767659
number of parsed urls = 143500, time = 14:55:04.538798
number of parsed urls = 143600, time = 14:55:40.047181
number of parsed urls = 143700, time = 14:56:12.667287
number of parsed urls = 143800, time = 14:56:53.097831
number of parsed urls = 143900, time = 14:57:28.424196
number of parsed urls = 144000, time = 14:58:01.394728
number of parsed urls = 144100, time = 14:58:38.787795
number of parsed urls = 144200, time = 14:59:13.936382
number of 

number of parsed urls = 157300, time = 16:17:50.777567
number of parsed urls = 157400, time = 16:18:27.047280
number of parsed urls = 157500, time = 16:19:01.253923
number of parsed urls = 157600, time = 16:19:39.238666
number of parsed urls = 157700, time = 16:20:13.556008
number of parsed urls = 157800, time = 16:20:50.854377
number of parsed urls = 157900, time = 16:21:27.686416
number of parsed urls = 158000, time = 16:22:03.908152
number of parsed urls = 158100, time = 16:22:37.918260
number of parsed urls = 158200, time = 16:23:13.321142
number of parsed urls = 158300, time = 16:23:48.034300
number of parsed urls = 158400, time = 16:24:23.682534
number of parsed urls = 158500, time = 16:25:00.173546
number of parsed urls = 158600, time = 16:25:35.855608
number of parsed urls = 158700, time = 16:26:08.527127
number of parsed urls = 158800, time = 16:26:40.473636
number of parsed urls = 158900, time = 16:27:16.123629
number of parsed urls = 159000, time = 16:27:51.297671
number of 

number of parsed urls = 171300, time = 17:40:34.618060
number of parsed urls = 171400, time = 17:41:07.344350
number of parsed urls = 171500, time = 17:41:42.603779


Он все скачал, но в какой-то момент повисло ядро и нормально результат посмотреть я не смог. Что ж распарсим файл и посмотрим что там.

In [2]:
urls_from_dict = set()

In [3]:
with open('storage.dict') as f:
    for line in f:
        urls_from_dict.add(line.strip().split()[0])

In [5]:
len(urls_from_dict)

172262

In [6]:
categ_num = len([url for url in urls_from_dict if "Category" in url])
categ_num

29582

In [7]:
documents = len(urls_from_dict) - categ_num
documents

142680

Итого, что мы получили. 142680 документов и 29582 страницы с категориями. Причем все документы уникальные так как ключ сравнения это netloc + path